system/lib/standalone/standalone.c

```diff
@@ -109,15 +202,247 @@ __attribute__((__weak__)) int __syscall_fcntl64(int fd, int cmd, ...) {
   return -ENOSYS;
 }
 
+void to_public_stat(const __wasi_filestat_t *in,
+                                  struct stat *out) {
+  *out = (struct stat){
+    .st_dev = in->dev,
+    .st_ino = in->ino,
+    .st_nlink = in->nlink,
+    .st_size = in->size,
+    .st_atim = __wasi_timestamp_to_timespec(in->atim),
+    .st_mtim = __wasi_timestamp_to_timespec(in->mtim),
+    .st_ctim = __wasi_timestamp_to_timespec(in->ctim),
+  };
+
+  // Convert file type to legacy types encoded in st_mode.
+  switch (in->filetype) {
+    case __WASI_FILETYPE_BLOCK_DEVICE:
+      out->st_mode |= S_IFBLK;
+      break;
+    case __WASI_FILETYPE_CHARACTER_DEVICE:
+      out->st_mode |= S_IFCHR;
+      break;
+    case __WASI_FILETYPE_DIRECTORY:
+      out->st_mode |= S_IFDIR;
+      break;
+    case __WASI_FILETYPE_REGULAR_FILE:
+      out->st_mode |= S_IFREG;
+      break;
+    case __WASI_FILETYPE_SOCKET_DGRAM:
+    case __WASI_FILETYPE_SOCKET_STREAM:
+      out->st_mode |= S_IFSOCK;
+      break;
+    case __WASI_FILETYPE_SYMBOLIC_LINK:
+      out->st_mode |= S_IFLNK;
+      break;
+  }
+}
+
 __attribute__((__weak__)) int __syscall_fstat64(int fd, intptr_t buf) {
```

Comment:
Elsewhere in this file it looks like we put the weak attribute on its own line.. maybe do that here to?

Reply:
It's not really consistent over the file, but I have changed it now where the method is implemented.